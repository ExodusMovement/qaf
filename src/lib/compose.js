import React from 'react';

const el = (Component, ...args) =>
  (typeof Component.type === 'function'
    ? React.cloneElement
    : React.createElement)(Component, ...args);

export default components => props => {
  const reducer = (memo, key) => allProps =>
    el(components[key], {
      children: childProps => memo({ ...allProps, [key]: childProps })
    });

  return Object.keys(components).reduceRight(
    reducer,
    props.children || props.render
  )();
};
