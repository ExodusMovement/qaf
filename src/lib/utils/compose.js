import { createElement } from 'react';

export default components => props => {
  const reducer = (memo, key) => allProps =>
    createElement(components[key], {}, childProps =>
      memo({ ...allProps, [key]: childProps })
    );

  return Object.keys(components).reduceRight(
    reducer,
    props.children || props.render
  )();
};
