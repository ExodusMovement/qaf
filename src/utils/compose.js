// @flow

import { createElement } from 'react';

export default (...components) => props =>
  components.reduceRight(
    (children, Component) => (...allProps) =>
      createElement(Component, {}, childProps =>
        children(...allProps.concat(childProps))
      ),
    props.children || props.render
  )();
