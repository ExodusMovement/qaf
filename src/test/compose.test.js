// @flow

/* eslint-disable react/prop-types */

import React, { Fragment } from 'react';

import { create as r } from 'react-test-renderer';

import compose from '../lib/utils/compose';

it('works', () => {
  const Foo = props => (
    <Fragment>{props.children({ youGot: 'fooed' })} </Fragment>
  );
  const Bar = props => (
    <Fragment>{props.children({ youGot: 'bared' })} </Fragment>
  );

  const Baz = () => (
    <Foo>
      {foo => (
        <Bar>
          {bar => (
            <Fragment>
              <Fragment>{foo.youGot}</Fragment>
              <Fragment>{bar.youGot}</Fragment>
            </Fragment>
          )}
        </Bar>
      )}
    </Foo>
  );

  const tree = r(<Baz />).toJSON();

  expect(tree.includes('fooed')).toBe(true);
  expect(tree.includes('bared')).toBe(true);

  const Composed = compose({
    foo: Foo,
    bar: Bar
  });

  const composed = r(
    <Composed
      render={({ foo, bar }) => (
        <Fragment>
          <Fragment>{foo.youGot}</Fragment>
          <Fragment>{bar.youGot}</Fragment>
        </Fragment>
      )}
    />
  ).toJSON();

  expect(composed).toEqual(tree);
});
