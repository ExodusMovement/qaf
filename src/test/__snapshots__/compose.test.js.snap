// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Array [
  "fooed",
  "bared",
  " ",
  " ",
]
`;

exports[`works 2`] = `
Array [
  "fooed",
  "bared",
  " ",
  " ",
]
`;
